src/main/kotlin/org/elm/workspace/ElmProject.kt

```diff
@@ -84,6 +87,20 @@ sealed class ElmProject(
     val allDirectDeps: Sequence<ElmPackageProject> =
             sequenceOf(dependencies, testDependencies).flatten()
 
+    private val sourceDirCache = AtomicReference<List<VirtualFile>>()
+
+    /**
+     * Returns [VirtualFile]s for each directory in [allSourceDirs].
+     */
+    val allSourceDirVirtualFiles: List<VirtualFile>
+        get() {
+            val cached = sourceDirCache.get()
+            if (cached?.all { it.isValid } == true) return cached
+            val dirs = allSourceDirs.mapNotNull { LocalFileSystem.getInstance().findFileByIoFile(it.toFile()) }.toList()
```

Comment:
This takes a couple of milliseconds to run, and `ElmCreateFileAction.isAvailable` is called frequently (e.g. every time you click in the project pane), so it's worth caching.

Reply:
This might no longer be necessary if you make the change suggested above. `ElmWorkspaceService.findProjectForFile()` uses an index (forked from intellij-rust) which should be pretty efficient. But you should double-check the actual performance difference.